### DISCUSSION TITLE
How to put database inside a struct?

### DISCUSSION THREAD

**Discussion by rix0rrr** [vincent-herlemont/native_db#71] [Q&A]
Hi,

I'm trying to use `native_db`, but I have a feeling I'm doing something wrong...

I'd like to hide the concern of how my database is implemented, so I'd like to hide the implementation details in a `struct` with some methods. But the combination of `DatabaseBuilder` and `Database` which borrows the builder is making this hard.

My first attempt was this, but the lifetime I have to pass into the `Database` leads to a parameter that I don't actually want to give. I want to pass a lifetime like `'self` or `'this` or something, but that doesn't exist:

```rust
pub struct Database<'a> {
    //            ^^^ it's not actually a parameter
    builder: DatabaseBuilder,
    db: native_db::Database<'a>,
}
```

It also becomes nearly impossible to implement the initializer for this struct:

```rust
    pub fn new() -> Result<Self> {
        let mut builder = DatabaseBuilder::new();
        builder.define::<Task>()?;
        Ok(Self {
            builder,
            db: builder.open("db.db")?,
            // ^^^^ error[E0597]: `builder` does not live long enough
        })
    }
```

I was ultimately somewhat able to make it work by doing contortions, by putting the `Database` into an `Option<>` and initializing it after construction.

```rust
pub struct Database<'a> {
    builder: DatabaseBuilder,
    db: Option<native_db::Database<'a>>,
}

impl<'a> Database<'a> {
    pub fn new() -> Result<Self> {
        let mut ret = Self {
            builder: DatabaseBuilder::new(),
            db: None,
        };
        ret.builder.define::<Task>()?;
        Ok(ret)
    }

    pub fn open(&'a mut self) -> Result<()> {
        //      ^^^^ this is the magic
        self.db = Some(self.builder.open("db.db")?);
        Ok(())
    }
}
```

This doesn't feel ideal, and everywhere I want to accept the database now I'm still stuck with that annoying lifetime parameter.

Another thing I figured I could do is just store the builder, and open and close the database on every mutation. But, I don't know. It's not clear to me that how it was designed to be used, and it intuitively feels wasteful to do that. 

So here's my question: how should I put a database connection inside a struct?

**Comment by vincent-herlemont** [comment]
@rix0rrr The issue you are encountering is attempting a self-reference in a `struct`, which is "possible" but very limited in Rust. In our case, it's not feasible because there is a custom drop in redb.

For a brief explanation: `Redb` is used in the backend and its definition, [`redb::TableDefinition`](https://github.com/cberner/redb/blob/9c28085c8465929b66b2ab10287b7ad087d2b7c7/src/db.rs#L130), has a `name: &'a str` field which is a reference. This forces us to store it in an external variable outside the [`Database`](https://github.com/vincent-herlemont/native_db/blob/7a1a4210a45195775a0eb8d883641a0c0b0fa80e/src/database.rs#L34-L39) struct.

:point_right: So, I would advise against trying to do this.

Instead, try to ensure that `DatabaseBuilder` lives as long as `Database`.

Feel free to post your new design.

**Comment by Gawdl3y** [comment]
This is essentially the same issue that I'm grappling with in #69. The requirement to keep the DatabaseBuilder around for the same life as Database is less than intuitive, I'd say, and actually nearly impossible to do in stateful applications without some hacky-feeling workarounds.

It looks like the creator of redb intends on removing as many lifetimes as possible from the public API for 2.0, based on a checklist item in cberner/redb#677 (and some of it has happened in cberner/redb#732). Hopefully that comes to fruition, as the lifetimes are making it tough to work with.
